### ISSUE TITLE
Usage FFMpegCodec Video Endpoint

### ISSUE THREAD

**Issue by thornzz** [sipsorcery/sipsorcery#366]
Hi,

I have tried to implement FFMpegCodec Video Endpoint to GetStartedVideo example project but i could not get video working.

I have changed only this lines. Is there any other thing to start video call? Video Form that show remote video is opening but its empty. I think it is not true usage of FFMpegCodec coz WindowsVideoEndpoint starting Webcam but FFmpegVideoEndpoint do nothing. I have checked FFMpegCodec usage at WebRTC examples but i need Sip server usage.

  var videoEndpointFFMepg = new FFmpegVideoEndPoint();

  MediaEndPoints mediaEndPoints = new MediaEndPoints
            {
                AudioSink = windowsAudioEndPoint,
                AudioSource = windowsAudioEndPoint,
                VideoSink = videoEndpointFFMepg,
                VideoSource = videoEndpointFFMepg,
            };

**Comment by sipsorcery** [comment]
What are you aiming to do?

The [SIPSorceryMedia.FFmpeg](https://github.com/sipsorcery/SIPSorceryMedia.FFmpeg) package doesn't have any integrations with audio or video devices wired up. For example you won't be able to use it to access a system webcam unless you want to enhance it and add those hooks in.

**Comment by thornzz** [comment]
There are some devices that only support H264 codec. Our windows application will wait incoming video call from that devices. So WindowsVideoEndpoint not working with them. I need H264 codec package for make a video call.

**Comment by sipsorcery** [comment]
You should be able to use a combination of the two video end points. Something along the lines of:

````
var ffmpegVideoEndPoint = new FFmpegVideoEndPoint();
ffmpegVideoEndPoint.RestrictFormats(format => format.Codec == VideoCodecsEnum.H264);
var windowsVideoEndPoint = new WindowsVideoEndPoint();

MediaEndPoints mediaEndPoints = new MediaEndPoints
{
	AudioSink = windowsAudioEndPoint,
	AudioSource = windowsAudioEndPoint,
	VideoSink = ffmpegVideoEndPoint,
	VideoSource = ffmpegVideoEndPoint,
};   

// Get local video samples from the Windows Video EndPoint.
windowsVideoEndPoint.OnVideoSourceRawSample => LocalVideoSampleHandler;
await windowsVideoEndPoint.StartVideo().ConfigureAwait(false);

ffmpegVideoEndPoint.OnVideoSinkDecodedSample += (byte[] bmp, uint width, uint height, int stride, VideoPixelFormatsEnum pixelFormat) => RemoteVideoHandler;

// When the call is answered connect the two end points.
windowsVideoEndPoint.OnVideoSourceRawSample += ffmpegVideoEndPoint.ExternalVideoSourceRawSample;

**Comment by thornzz** [comment]
I have tried this code but webcam that application is working on not started and no video remote or local.

This line not started webcam => windowsVideoEndPoint.StartVideo().ConfigureAwait(false);

**Comment by sipsorcery** [comment]
That's the same line as in the existing {GetStartedVideo demo](https://github.com/sipsorcery/sipsorcery/blob/master/examples/SIPExamples/GetStartedVideo/Program.cs).

The change is you're now adding a new `FFmpegVideoEndPoint` and using it in the `VoIPMediaSession`.

**Comment by thornzz** [comment]
Im getting this errors after call started Voice is working but no video.

[22:03:08 WRN] RTCP session for local ssrc 66027432 has not had any activity for over 30 seconds.
[22:03:20 WRN] Could not match an RTCP packet against any SSRC's in the session.
[22:03:20 WRN] Could not match an RTCP packet against any SSRC's in the session.
[22:03:20 WRN] Could not match an RTCP packet against any SSRC's in the session.
[22:03:25 WRN] Could not match an RTCP packet against any SSRC's in the session.

**Comment by sipsorcery** [comment]
I refactored the media assemblies a little bit to make this scenario easier. The code required is now:

````
var windowsVideoEndPoint = new WindowsVideoEndPoint(new FFmpegVideoEncoder());
windowsVideoEndPoint.RestrictFormats(format => format.Codec == VideoCodecsEnum.H264);
````

BUT I overlooked the critical lack of `H264` depacketisation not being implemented in `RTPSession`. That means it while it's possible to send H264 video streams it's not possible to extract the received `H264` frames from the `RTP` packets to decode them. I've added #378 to capture the enhancement. In the meantime it's not possible to establish a SIP based video call with `H264`. The only supported option is `VP8`.
